File: preprocessing.py
import numpy as np
import scipy.sparse


def load_lastfm(path="./lastfm-dataset-360K/usersha1-artmbid-artname-plays.tsv"):
    def index(x, name):
        i = x.get(name)
        if i is None:
            x[name] = i = len(x)
        return i

    n_lines = 0
    with open(path, "r") as inp:
        for line in inp:
            n_lines += 1
    data = np.empty((n_lines, 3), dtype=int)
    users = {}
    items = {}
    with open(path, "r") as inp:
        for i, line in enumerate(inp):
            cells = line.split('\t')
            data[i, 0] = index(users, cells[0])
            data[i, 1] = index(items, cells[1])
            data[i, 2] = int(cells[3])
    return data


lastfm = load_lastfm()
lastfm = scipy.sparse.csr_matrix(
    (lastfm[:, 2], (lastfm[:, 0], lastfm[:, 1])), dtype=np.float32)
scipy.sparse.save_npz('lastfm.npz', lastfm)

File: preprocessing.jl
using DelimitedFiles
using Serialization
using SparseArrays


function load_lastfm(path="./lastfm-dataset-360K/usersha1-artmbid-artname-plays.tsv")
    function index!(dict, name)
        i = get(dict, name, 0)
        if i == 0
            dict[name] = i = length(dict) + 1
        end
        return i
    end

    data = Array{Int}(undef, countlines(open(path)), 3)
    users = Dict{String, Int}()
    items = Dict{String, Int}()
    for (i, line) in enumerate(eachline(path))
        cells = split(line, '\t')
        data[i, 1] = index!(users, cells[1])
        data[i, 2] = index!(items, cells[2])
        data[i, 3] = parse(Int, cells[4])
    end

    return data
end

lastfm = load_lastfm()
lastfm = sparse(view(lastfm, :, 2),
                view(lastfm, :, 1),
                view(lastfm, :, 3))
serialize("lastfm.jlser", lastfm)
